### DISCUSSION TITLE
AfterSave Hook

### DISCUSSION THREAD

**Discussion by ariesdg** [filamentphp/filament#6850] [Help]
Hi,

Newbie here, I was trying to create an event right after the record is saved. However, the afterSave() hook is intermittently working. Like randomly it's working and sometimes it is not. I think the Laravel wasChanged() method causing it, any advice on the best approach? I tried using an observer to observe any changes in the model and fire an event, but seems the code is hidden and cannot be read by future dev. 

```
protected function afterSave(): void
  {
     if($this->record->wasChanged('attribute_id')
     {
         //my code here
     }
  }
```

**Comment by danharrin** [comment]
Are you changing the `attribute_id` to a different value every time?

**Reply by ariesdg** [reply]
Yes, I need to fire an event every time the attribute changes. And the event will depend on what is the value of the attribute

**Reply by danharrin** [reply]
Without being able to replicate the problem consistently I cant help you

**Reply by ariesdg** [reply]
Sorry, I had to rewrite my previous code.

I have a Ticket table and has a column of `ticket_status_id`. The Ticket has a relation of HasMany on `TicketWorkProgress` model. Every time the status changes into id` of 2 (which is in progress), it will add a record on the `TicketWorkProgress`. And every time the status changes on any `id` it will update the TicketWorkProgress column `completed_at`. 

Here is my resource
<img width="230" alt="image" src="https://github.com/filamentphp/filament/assets/51257627/879e0a64-ff0b-4322-b98e-52afc1ca5053">

Here is my hook function on ticket edit page

```
protected function afterSave(): void
{
     if($this->record->wasChanged('ticket_status_id')
     {
           if($this->record->ticket_status_id == 2){
                   $workProgress = new TicketWorkProgress();
                   $workProgress->initiated_by = auth()->id();
                   $workProgress->modified_by = auth()->id();
                   $workProgress->ticket_id = $this->record->id;
                   $workProgress->save();
            }else{
                     $workDuration = $this->record->created_at->diffInMinutes(Carbon::now());
                    
                    if($workDuration <= 0){
                         $workDuration = 1;
                    }
                   if($this->record->ticketWorkProgress->count() > 0){
                      $currentWork = $this->record->ticketWorkProgress()->whereNull('completed_at')->orderBy('created_at', 'desc')->first();
            
                        if($currentWork){
                
                             $currentWork->work_duration = $workDuration;
                             $currentWork->completed_at = now();
                             $currentWork->modified_by = auth()->id();
                             $currentWork->save();
                         }
                   }
            }
          
     }
}
```

**Reply by ArtMin96** [reply]
Do you want this? [Customizing the saving process](https://filamentphp.com/docs/2.x/admin/resources/editing-records#customizing-the-saving-process)
